### PR TITLE
refactor(rust): Use Zero trait for cumulative function state initialization

### PR DESCRIPTION
**Pull request by WillAyd**
This also makes it so that these functions have the same signatures as `det_min` and `det_max` within the same module

### CODE REVIEW

**Comment by ritchie46** [comment]
Can you look at the tests?

**Comment by WillAyd** [comment]
Hmm sorry should have done that locally first. I guess this doesn't work with `det_prod` because using 0 as an initial value doesn't work with the multiplication.

Not sure this is worth standalone for `det_sum` - I'll see if it makes sense to do in a larger change rather than this one off

**Comment by ritchie46** [comment]
For multiplication you can use 1

**Comment by WillAyd** [comment]
Fair enough. Just pushed that up

File: polars/polars-core/src/chunked_array/ops/cum_agg.rs [hunk 1]
```diff
@@ -36,37 +36,29 @@ where
...
+            *state = *state + v;
+            Some(Some(*state))
         }
-        (_, None) => Some(None),
+        None => Some(None),
     }
 }
 
-fn det_prod<T>(state: &mut Option<T>, v: Option<T>) -> Option<Option<T>>
+fn det_prod<T>(state: &mut T, v: Option<T>) -> Option<Option<T>>
```
Comment (ritchie46) [inline polars/polars-core/src/chunked_array/ops/cum_agg.rs:52]: I think now should return `Option<T>` as the identity isn't `None` anymore, but `0` or `1`.

File: polars/polars-core/src/chunked_array/ops/cum_agg.rs [hunk 2]
```diff
@@ -36,37 +36,29 @@ where
...
-        (None, Some(v)) => {
-            *state = Some(v);
-            Some(*state)
+    match v {
+        Some(v) => {
+            *state = *state * v;
+            Some(Some(*state))
         }
-        (_, None) => Some(None),
+        None => Some(None),
```
Comment (ritchie46) [inline polars/polars-core/src/chunked_array/ops/cum_agg.rs:61]: And then we should always return `Some(*state)`.

File: polars/polars-core/src/chunked_array/ops/cum_agg.rs (hunk 1 above)
Comment (WillAyd) [reply polars/polars-core/src/chunked_array/ops/cum_agg.rs:52]: First attempt at this doesn't play nicely with the `collect_trusted` iter.

```sh
error[E0271]: expected `Box<dyn PolarsIterator<Item = Option<<T as PolarsNumericType>::Native>>>` to be an iterator that yields `<T as PolarsNumericType>::Native`, but it yields `Option<<T as PolarsNumericType>::Native>`
  --> polars/polars-core/src/chunked_array/ops/cum_agg.rs:94:59
   |
94 |             false => self.into_iter().scan(init, det_sum).collect_trusted(),
   |                                                           ^^^^^^^^^^^^^^^ expected associated type, found `Option<<... as PolarsNumericType>::Native>`
   |
   = note: expected associated type `<T as datatypes::PolarsNumericType>::Native`
                         found enum `std::option::Option<<T as datatypes::PolarsNumericType>::Native>`
   = help: consider constraining the associated type `<T as datatypes::PolarsNumericType>::Native` to `std::option::Option<<T as datatypes::PolarsNumericType>::Native>` or calling a method that returns `<T as datatypes::PolarsNumericType>::Native`
   = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
   = note: required for `Scan<Box<dyn PolarsIterator<Item = Option<<T as PolarsNumericType>::Native>>>, ..., ...>` to implement `polars_arrow::trusted_len::TrustedLen`
   = note: the full type name has been written to '/home/willayd/clones/polars/target/debug/deps/polars_core-f27f8b388d927e0a.long-type-3305089848937386963.txt'
```

Need to dig a bit further into how that is set up

**Pull request closed without merging**